File: ashok/platformio/EE2010_52/ee2010_52.py
def get_kmap_input():
    print("Enter values for the 2x4 K-map (0 or 1 only):")
    print("Follow this minterm order: m0 m1 m3 m2 m4 m5 m7 m6")
    minterm_order = [0, 1, 3, 2, 4, 5, 7, 6]
    kmap = [0] * 8
    for idx in minterm_order:
        val = int(input(f"Enter value for m{idx}: "))
        kmap[idx] = val
    return kmap

def print_kmap(kmap):
    print("\nK-map (X vs YZ):")
    print("        YZ")
    print("       00  01  11  10")
    print("X=0:  {}   {}   {}   {}".format(kmap[0], kmap[1], kmap[3], kmap[2]))
    print("X=1:  {}   {}   {}   {}".format(kmap[4], kmap[5], kmap[7], kmap[6]))

def get_expression(kmap):
    terms = []
    labels = [
        "X'Y'Z'", "X'Y'Z", "X'YZ'", "X'YZ", 
        "XY'Z'", "XY'Z", "XYZ'", "XYZ"
    ]
    for i in range(8):
        if kmap[i] == 1:
            terms.append(labels[i])
    
    if not terms:
        return "0"
    elif len(terms) == 8:
        return "1"
    else:
        return " + ".join(terms)

def main():
    kmap = get_kmap_input()
    print_kmap(kmap)
    expression = get_expression(kmap)
    print("\nSimplified Boolean Expression (Sum of Minterms):")
    print("F(X, Y, Z) =", expression)

if __name__ == "__main__":
    main()

File: ashok/gate questions/c/EE2010_52.c
#include <stdio.h>
#include <string.h>

void get_kmap_input(int kmap[8]) {
    int minterm_order[8] = {0, 1, 3, 2, 4, 5, 7, 6};
    int val;

    printf("Enter values for the 2x4 K-map (0 or 1 only):\n");
    printf("Follow this minterm order: m0 m1 m3 m2 m4 m5 m7 m6\n");

    for (int i = 0; i < 8; i++) {
        int idx = minterm_order[i];
        printf("Enter value for m%d: ", idx);
        scanf("%d", &val);
        if (val != 0 && val != 1) {
            printf("Invalid input. Only 0 or 1 allowed.\n");
            i--;  // redo input for this index
        } else {
            kmap[idx] = val;
        }
    }
}

void print_kmap(int kmap[8]) {
    printf("\nK-map (X vs YZ):\n");
    printf("        YZ\n");
    printf("       00  01  11  10\n");
    printf("X=0:   %d   %d   %d   %d\n", kmap[0], kmap[1], kmap[3], kmap[2]);
    printf("X=1:   %d   %d   %d   %d\n", kmap[4], kmap[5], kmap[7], kmap[6]);
}

void get_expression(int kmap[8], char* expression) {
    const char* labels[8] = {
        "X'Y'Z'", "X'Y'Z", "X'YZ'", "X'YZ",
        "XY'Z'", "XY'Z", "XYZ'", "XYZ"
    };

    int first = 1;
    strcpy(expression, "");

    int all_zero = 1, all_one = 1;
    for (int i = 0; i < 8; i++) {
        if (kmap[i] != 0) all_zero = 0;
        if (kmap[i] != 1) all_one = 0;
    }

    if (all_zero) {
        strcpy(expression, "0");
        return;
    }
    if (all_one) {
        strcpy(expression, "1");
        return;
    }

    for (int i = 0; i < 8; i++) {
        if (kmap[i] == 1) {
            if (!first) strcat(expression, " + ");
            strcat(expression, labels[i]);
            first = 0;
        }
    }
}

int main() {
    int kmap[8];
    char expression[256];

    get_kmap_input(kmap);
    print_kmap(kmap);
    get_expression(kmap, expression);

    printf("\nSimplified Boolean Expression (Sum of Minterms):\n");
    printf("F(X, Y, Z) = %s\n", expression);

    return 0;
}

